libraries/AP_Math/matrix_alg.cpp

```diff
@@ -149,71 +126,71 @@ static void mat_back_sub(const T *U, T *out, uint16_t n)
  *    @param     n,           dimension of matrix
  */
 template<typename T>
-static void mat_LU_decompose(const T* A, T* L, T* U, T *P, uint16_t n)
+static void mat_LU_decompose(const T* A, T* L, T* U, T *P, T *APrime, uint16_t n)
 {
     memset(L,0,n*n*sizeof(T));
     memset(U,0,n*n*sizeof(T));
     memset(P,0,n*n*sizeof(T));
     mat_pivot(A,P,n);
 
-    T *APrime = matrix_multiply(P,A,n);
+    mat_mul(P, A, APrime, n);
     for(uint16_t i = 0; i < n; i++) {
         L[i*n + i] = 1;
     }
     for(uint16_t i = 0; i < n; i++) {
         for(uint16_t j = 0; j < n; j++) {
-            if(j <= i) {    
+            if(j <= i) {
                 U[j*n + i] = APrime[j*n + i];
                 for(uint16_t k = 0; k < j; k++) {
-                    U[j*n + i] -= L[j*n + k] * U[k*n + i]; 
+                    U[j*n + i] -= L[j*n + k] * U[k*n + i];
                 }
             }
             if(j >= i) {
                 L[j*n + i] = APrime[j*n + i];
                 for(uint16_t k = 0; k < i; k++) {
-                    L[j*n + i] -= L[j*n + k] * U[k*n + i]; 
+                    L[j*n + i] -= L[j*n + k] * U[k*n + i];
                 }
                 L[j*n + i] /= U[i*n + i];
             }
         }
     }
-    delete[] APrime;
 }
 
+
 /*
  *    matrix inverse code for any square matrix using LU decomposition
  *    inv = inv(U)*inv(L)*P, where L and U are triagular matrices and P the pivot matrix
  *    ref: http://www.cl.cam.ac.uk/teaching/1314/NumMethods/supporting/mcmaster-kiruba-ludecomp.pdf
  *    @param     m,           input 4x4 matrix
- *    @param     inv,      Output inverted 4x4 matrix
- *    @param     n,           dimension of square matrix
+ *    @param     inv,         Output inverted 4x4 matrix
+ *    @tparam    n,           dimension of square matrix
+ *    @param     buf,         buffer handler for temp matrices
  *    @returns                false = matrix is Singular, true = matrix inversion successful
  */
 template<typename T>
-static bool mat_inverseN(const T* A, T* inv, uint16_t n)
+static bool mat_inverseN(const T* A, T* inv, const uint16_t n, MatInvBuffer<T> &buf)
 {
-    T *L, *U, *P;
-    bool ret = true;
-    L = NEW_NOTHROW T[n*n];
-    U = NEW_NOTHROW T[n*n];
-    P = NEW_NOTHROW T[n*n];
-    mat_LU_decompose(A,L,U,P,n);
+    T *L      = buf.L();
```

Comment:
i think it would be a lot simpler to declare a structure with the elements it needs, then allocation that with a single NEW_NOTHROW, and return false on failure
I'm just concerned this PR spreads the logic of the mat inverse across 3 files, with entanglement between them, with the U(), P() etc methods in AP_Math.h, which are of specific sizes needed by this code